src/cmd/builtin/proxy.c

```diff
@@ -289,6 +289,56 @@ static void proxy_command_destroy_usock_and_router (struct proxy_command *ctx)
     ctx->router = NULL;
 }
 
+/* Attempt to reconnect to broker.  If successful, wait for for broker to
+ * reach RUN state.  This function exits on all fatal errors, so it is safe
+ * to call again on failure.
+ *
+ * Notes
+ * - error callback is protected against reentry
+ * - send/recv within the callback can fail as though no callback is registered
+ * - state-machine.wait RPC fails if broker is past the RUN state
+ * - it is safe to call flux_reconnect(3) if already connected
+ */
+static int try_reconnect (flux_t *h, struct router *router)
+{
+    flux_future_t *f;
+    int rc = -1;
+
+    if (flux_reconnect (h) < 0) {
+        if (errno == ENOSYS)
+            log_msg_exit ("reconnect not implemented by connector");
+        return -1;
+    }
+    if (!(f = flux_rpc (h, "state-machine.wait", NULL, FLUX_NODEID_ANY, 0))) {
+        log_err ("could not send state-machine.wait RPC");
+        goto done;
+    }
+    if (flux_rpc_get (f, NULL) < 0) {
+        log_msg ("state-machine.wait: %s", future_strerror (f, errno));
+        goto done;
+    }
```

Comment:
I'm confused by the comment above:

> ` state-machine.wait RPC fails if broker is past the RUN state`

and the code here which skips `router_renew()` when `state-machine.wait` fails.

If a broker restarts and `flux-proxy` attempts a reconnect, but by the broker is already in RUN state by the time the `state-machine.wait` RPC is sent, shouldn' t the `router_renew()` call still be made?

I'm probably just missing something subtle.

Also, is `router_renew()` only useful for `flux-proxy`? It would be nice if the shell didn't have to re-establish its own subscriptions and service registrations after a reconnect, but it isn't a router so cannot use `router_renew()` AFAICS. Probably it is actually impossible to handle this case for clients connected using the local connector, but I thought I'd ask for clarity's sake.

Reply:
I'll add some more inline commentary as this is not very clear, but the idea was that the `state-machine.wait` RPC would fail if the broker is _past_ the RUN state, e.g. it is shutting down.   The reconnect logic then does its timed retry, giving the broker a chance to be restarted by systemd.

Reply:
Ugh, sorry. My fault. I had read that as "RUN state or beyond".

Reply:
Yeah, `router_renew()` is only for librouter users like `flux-proxy`.

It might be possible to add a shell plugin that acts as a router for processes spawned by the shell, if that's useful.  This would not take much code if we wanted to try it.

I hadn't considered whether we could somehow use librouter from the shell thread itself.   librouter is meant to be used to build a component that is inserted in the messsage path between clients and the broker (or another router).   Might be worth thinking about that.

Reply:
This is a very intriguing idea!

We already have some messages that shell internal only. For example, the rank 0 output plugin sends messages to itself. It would be neat if that could be handled internal to the shell somehow.